underworld/libUnderworld/StgFEM/Utils/src/SemiLagrangianIntegrator.c

```diff
@@ -541,62 +556,88 @@ void BicubicInterpolator( FeVariable* feVariable, double* position, double* delt
-   double				delta[3];
-   unsigned			nInc;
-   unsigned*			inc;
+   FeMesh*   feMesh   = variableField->feMesh;
+   unsigned  meshSize = Mesh_GetLocalSize( feMesh, MT_VERTEX );
+   unsigned  nDims    = Mesh_GetDimSize( feMesh );
+   Grid**    nodegrid = (Grid**) Mesh_GetExtension( feMesh, Grid*,  feMesh->vertGridId );
+   unsigned* nNodes   = Grid_GetSizes( *nodegrid );
+
+   unsigned  node_I;
+   double    var[3], x_i[3], delta[3], minLength, *x_0;
+   Bool      result;
+   
+   Mesh_GetMinimumSeparation( feMesh, &minLength, delta );
 
-   FeMesh_GetElementNodes( variableField->feMesh, 0, variableField->inc );
+#if 0
+   FeMesh_GetElementNodes( feMesh, 0, variableField->inc );
    nInc = IArray_GetSize( variableField->inc );
-   inc = IArray_GetPtr( variableField->inc );
 
-   delta[0] = Mesh_GetVertex( feMesh, inc[1] )[0] - Mesh_GetVertex( feMesh, inc[0] )[0];
    if( nInc % 3 == 0 ) /* quadratic elements */ {
-      delta[1] = Mesh_GetVertex( feMesh, inc[3] )[1] - Mesh_GetVertex( feMesh, inc[0] )[1];
-      if( nDims == 3 )
-         delta[2] = Mesh_GetVertex( feMesh, inc[9] )[2] - Mesh_GetVertex( feMesh, inc[0] )[2];
       for( dim_I = 0; dim_I < nDims; dim_I++ )
          nNodes[dim_I] = 2 * sizes[dim_I] + 1;
-   }
-   else {
-      delta[1] = Mesh_GetVertex( feMesh, inc[2] )[1] - Mesh_GetVertex( feMesh, inc[0] )[1];
-      if( nDims == 3 )
-         delta[2] = Mesh_GetVertex( feMesh, inc[4] )[2] - Mesh_GetVertex( feMesh, inc[0] )[2];
-
+   } else {
       for( dim_I = 0; dim_I < nDims; dim_I++ )
          nNodes[dim_I] = sizes[dim_I] + 1;
    }
+#endif
 
    FeVariable_SyncShadowValues( velocityField );
    FeVariable_SyncShadowValues( variableField );
-
+   
    /* assume that the variable mesh is the same as the velocity mesh */
    for( node_I = 0; node_I < meshSize; node_I++ ) {
-     
-      /* find the position back in time (u*) */
-      IntegrateRungeKutta( velocityField, dt, Mesh_GetVertex( feMesh, node_I ), position );
+      FeVariable_GetValueAtNode( variableField, node_I, var );
 
-      /* create a bicubic interpolation of variableField at u* */
-      BicubicInterpolator( variableField, position, delta, nNodes, var );
+      /* find the position back in time (u*), x_i */
+      x_0 = Mesh_GetVertex(feMesh, node_I);
+      IntegrateRungeKutta( velocityField, dt, x_0, x_i );
+
+      /* if the new x_i is "close" to original node, don't Bicubuic Interpolate, take original node value */
+      if( !SemiLagrangianIntegrator_PointsAreClose(x_i, x_0, nDims, 0, 1e-6*minLength) ) {
+        // if BicubicInerpolator returns false in means x_i was not in domain space
+        result = BicubicInterpolator(variableField, x_i, delta, nNodes, var);
+
+        // Commenting this out
+         if(result == False) {
+          printf("Interpolation failed, because position is not it domain space, using the original value on the vertex");
+         }
+        //
+      }
 
       FeVariable_SetValueAtNode( varStarField, node_I, var );
```

Comment:
If I'm understanding this correctly, `var` is only set if `result==True`. Otherwise, the result retrieved at `:620` is used. 

If that's the case, I think this isn't obvious to someone reading the code.  Better to be explicit:

```c
if (result==False)
{
// fallback to using nodal value
FeVariable_GetValueAtNode( variableField, node_I, var )
}
FeVariable_SetValueAtNode( varStarField, node_I, var );
```

Reply:
This also saves a memory access, which can make a significant difference for cache usage.